### ISSUE TITLE
prebuilt node binary is always downloaded even after the second run

### ISSUE THREAD

**Issue by lucidfrontier45** [RobertCraigie/pyright-python#153]
context https://github.com/microsoft/pyright/issues/4656

## Environment

- OS: Windows 11, WSL2, Debian 11
- Python: 3.11.1

## Description

Since 1.1.293 pyright-python always downloads and install prebuilt node binary even after the second run.

## Expected behaviour

It won't re-download node binary after the second run.

## How to reproduce

1. ensure `node` is not in `PATH`
1. create a venv `python3.11 -m venv --upgrade-deps venv
2. install pyright-python newer than 1.1.293 `./venv/bin/pip install pyright=1.1.294`
3. run pyright twice `./venv/bin/pyright` x2 

log of 1.1.292
```
du at du-thinkbook in ~/workspace/pyright_test
$ time ./venv/bin/pyright --version
WARNING: there is a new pyright version available (v1.1.292 -> v1.1.294). 
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

pyright 1.1.292
./venv/bin/pyright --version  1.55s user 0.16s system 76% cpu 2.246 total

du at du-thinkbook in ~/workspace/pyright_test
$ time ./venv/bin/pyright --version
WARNING: there is a new pyright version available (v1.1.292 -> v1.1.294).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

pyright 1.1.292
./venv/bin/pyright --version  1.50s user 0.22s system 103% cpu 1.669 total
```

log of 1.1.294
```
du at du-thinkbook in ~/workspace/pyright_test
$ time ./venv/bin/pyright --version
/home/du/workspace/pyright_test/venv/lib/python3.11/site-packages/nodeenv.py:26: DeprecationWarning: 'pipes' is deprecated and slated for removal in Python 3.13
  import pipes
 * Install prebuilt node (19.6.1) ..... done.
pyright 1.1.294
./venv/bin/pyright --version  4.17s user 0.97s system 38% cpu 13.198 total

du at du-thinkbook in ~/workspace/pyright_test
$ time ./venv/bin/pyright --version
/home/du/workspace/pyright_test/venv/lib/python3.11/site-packages/nodeenv.py:26: DeprecationWarning: 'pipes' is deprecated and slated for removal in Python 3.13
  import pipes
 * Install prebuilt node (19.6.1) ..... done.
pyright 1.1.294
./venv/bin/pyright --version  4.40s user 1.10s system 47% cpu 11.487 total
```

**Comment by tarvaina** [comment]
I also have this problem. It seems that 1.1.292 works as desired but 1.1.293 always redownloads node.

**Comment by tarvaina** [comment]
The cause of the problem is that here

https://github.com/RobertCraigie/pyright-python/blob/1087381cd7dc79d31f1d8bb8d732839d72cc4c9f/pyright/_utils.py#L4-L10

The builtin `hash(user)` returns different result for the same string in different runs. You probably want to use `hashlib.md5(user.encode()).hexdigest()` or something like it instead.

**Comment by RobertCraigie** [comment]
Ugh I'm really sorry I didn't catch this when initially testing that change. I think I must've accidentally had `PYTHONHASHSEED` set when testing this myself :/

Thanks for figuring it out!

I had actually inadvertently fixed this issue already by moving away from the badly designed temporary cache directory anyway in #136 but it hasn't been merged yet due to a failing windows test. After that PR is merged the cache directory will now use the actual user cache directory instead of a temporary directory mimicking user-specific caches.

So now resolving the root cache directory follows this pattern:
- Checks the `PYRIGHT_PYTHON_CACHE_DIR` environment variable
- Checks the `XDG_CACHE_HOME` environment variable
- Then if neither of the above are set, it defaults to `~/.cache`

I think this is much better behaviour as the previous solution meant that the nodeenv directory could be cleaned by the os more frequently than you'd probably want and it means we don't need the hacky logic to create user specific temporary directories.

**Comment by RobertCraigie** [comment]
The aforementioned change to the cache directory and the fix for this bug have now been released in `v1.1.296`.
